kotlinx-coroutines-core/common/src/selects/Select.kt

```diff
@@ -52,606 +98,728 @@ public interface SelectBuilder<in R> {
+        this.internalResult = NO_RESULT
+        this.clauses = null
+    }
 
-        private fun prepareSelectOp(): Any? {
-            impl._state.loop { state ->
-                when {
-                    state === this -> return null // already in progress
-                    state is OpDescriptor -> state.perform(impl) // help
-                    state === NOT_SELECTED -> {
-                        if (impl._state.compareAndSet(NOT_SELECTED, this))
-                            return null // success
-                    }
-                    else -> return ALREADY_SELECTED
-                }
-            }
+    /**
+     * Each `select` clause is internally represented with a [ClauseData] instance.
+      */
+    protected class ClauseData<R>(
+        @JvmField val clauseObject: Any, // the object of this `select` clause: Channel, Mutex, Job, ...
+        private val regFunc: RegistrationFunction,
+        private val processResFunc: ProcessResultFunction,
+        private val param: Any?, // the user-specified param
+        private val block: Any, // the user-specified block, which should be called if this clause becomes selected
+        @JvmField val onCancellationConstructor: OnCancellationConstructor?,
+        @JvmField var disposableHandle: DisposableHandle? = null
+    ) {
+        /**
+         * Tries to register the specified [select] instance in [clauseObject] and check
+         * whether the registration succeeded or a rendezvous has happened during the registration.
+         * This function returns `true` if this [select] is successfully registered and
+         * is _waiting_ for a rendezvous, or `false` when this clause becomes
+         * selected during registration.
+         *
+         * For example, the [Channel.onReceive] clause registration
+         * on a non-empty channel retrieves the first element and completes
+         * the corresponding [select] via [SelectInstance.selectInRegistrationPhase].
+         */
+        fun tryRegisterAsWaiter(select: SelectImplementation<R>): Boolean {
+            assert { select.inRegistrationPhase || select.isCancelled }
+            assert { select.internalResult === NO_RESULT }
+            regFunc(clauseObject, select, param)
+            return select.internalResult === NO_RESULT
         }
 
-        // reverts the change done by prepareSelectedOp
-        private fun undoPrepare() {
-            impl._state.compareAndSet(this, NOT_SELECTED)
+        /**
+         * Processes the internal result provided via either
+         * [SelectInstance.selectInRegistrationPhase] or
+         * [SelectInstance.trySelect] and returns an argument
+         * for the user-specified [block].
+         *
+         * Importantly, this function may throw an exception
+         * (e.g., when the channel is closed in [Channel.onSend], the
+         * corresponding [ProcessResultFunction] is bound to fail).
+         */
+        fun processResult(result: Any?) = try {
```

Comment:
In order to make a proper recovery, this function should be `suspend` and it should use `recoverAndThrow`.

`recoverStackTrace(e)` is basically meaningless and was introduced for a very narrow use-case